MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BoundingBox.java

```diff
@@ -0,0 +1,124 @@
+package com.google.android.libraries.motionphotoreader;
+
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+
+import static com.google.android.libraries.motionphotoreader.Constants.EPS;
+
+/**
+ * A BoundingBox object represents the bounding box for a quadrilateral, i.e. the box enclosing the
+ * quadrilateral with the smallest measure.
+ */
+public class BoundingBox {
+
+    private static final String TAG = "BoundingBox";
+
+    /**
+     * The extreme coordinates of the bounding box.
+     */
+    public final float xMin;
+    public final float yMin;
+    public final float xMax;
+    public final float yMax;
+
+    /**
+     * A pseudo-measurement of how well the bounding box approximates the quadrilateral. This is
+     * computed by taking the maximum distance from a vertex of the quadrilateral to a neighboring
+     * vertex of the bounding box. Thus, the error lies in the range [0, max(W, H) / 2], where
+     * W is the width of the bounding box and H is the height of the bounding box.
+     */
+    private final float error;
+
+    /**
+     * Constructor for a bounding box using the bounding box coordinates.
+     * @param xMin The x-coordinate of the lower-left vertex of the bounding box.
+     * @param yMin The y-coordinate of the lower=left vertex of the bounding box.
+     * @param xMax The x-coordinate of the upper-right vertex of the bounding box.
+     * @param yMax The y-coordinate of the upper-right vertex of the bounding box.
+     */
+    public BoundingBox(float xMin, float yMin, float xMax, float yMax) {
+        this.xMin = xMin;
+        this.yMin = yMin;
+        this.xMax = xMax;
+        this.yMax = yMax;
+        error = 0.0f;
+    }
+
+    /**
+     * Constructor for a bounding box using a given quadrilateral.
+     * The arguments a, b, c, d are all homogeneous 2D coordinates (x, y, 1) of the vertices of the
```

Comment:
If the third coordinate is always 1, why pass it? Furthermore, it seems that the coordinates are not treated as homogenous inside the code (that is, you are not dividing the first and second coordinate by the third).

Reply:
Most of the coordinates I use in the rest of the API are homogeneous coordinates with z=1 (since they have to be multiplied by a bunch of 3x3 homography matrices) so I thought it would be easiest to just have a constructor that specified the coordinates are homogeneous. The homography matrices also never change the value of z so I never divided the z-coordinate out anywhere. There's nothing in the BoundingBox code that needs the extra z-dimension though, so should I just remove this Javadoc line?

Also, I could pass in 2-dimensional coordinates but I would have to create another 3-dimensional coordinate for each point when I multiply them by homography matrices again, so I figured I should just use the 3-dimensional coordinates everywhere. Should I still just switch over to non-homogeneous coordinates in this case?

Reply:
I see. If we had time I would probably still recommend to change it (because you are essentially passing useless information that might even cause errors in the future if people start treating it like actual homogeneous coordinates).

One example of problems that might arise is that you are manipulating these homogeneous coordinates with 3x3 homographies, which might of course change the third coordinate if perspective-like transforms are applied. These resulting homogeneous coordinates will not have w = 1 anymore, and therefore will cause issues with functions such as the one we are commenting on.

One solution could be to have 2D (x, y) coordinates everywhere, and refactor your homography multiplication code to take a 2D vector and return a 2D vector (if need be, divide by the third coordinate as a last step in that function).

But for now, as I said, since we're out of time, we can keep it like that.

Another solution is to treat them like real homogeneous coordinates everywhere.

Reply:
Ok I see--I'll make a note of this in my final documentation. Thanks!